### PR TITLE
Added bulk import of whitelist, admin and banlist

### PR DESCRIPTION
**Pull request by jake-87**

### CODE REVIEW

**Comment by jake-87** [comment]
At the moment I have the ban reason as "bulk import"; should it also be imported from the file? It would complicate the creation of a ban list, but could be useful to a banned user.

**Comment by Hornwitser** [comment]
Usernames don't have spaces in them so you could separate the name and reason with that.

**Comment by jake-87** [comment]
@Hornwitser Should be ready for review.

**Comment by Hornwitser** [comment]
This is missing a way to do it through the web interface.

**Review by Danielv123 — APPROVED** [review]
This looks good to me. We should absolutely have the same functionality available in the web interface, whether in this PR or its own.

File: packages/ctl/ctl.js [hunk 1]
```diff
@@ -903,6 +903,29 @@ userCommands.add(new libCommand.Command({
...
+		yargs.positional("file", { describe: "File to import from", type: "string" });
+		yargs.options({
+			"create": { describe: "Create user if it does not exist", nargs: 0, type: "boolean", default: false },
+		});
+	}],
+	handler: async function(args, control) {
+		let data = await fs.readFile(args.file, "utf8");
+		let names = data.split("\n");
+		for (let name in names) {
+			if (name in names) {
```
Comment (Danielv123) [inline packages/ctl/ctl.js:917]: What is the point of this though?

**Comment by jake-87** [comment]
I’d be happy add the webui in this PR, however it may be a few days.

File: packages/ctl/ctl.js (hunk 1 above)
Comment (jake-87) [reply packages/ctl/ctl.js:917]: Not sure myself. eslint has spit out errors in the past if it’s not there, so I preemptively added it. Unfortunately I can’t test if it’s needed right now, but if it’s not, i’d be happy to remove it.

Comment (Hornwitser) [reply packages/ctl/ctl.js:917]: `for (... in ...)` loops are an error more often than not, you shouldn't use them.

Comment (jake-87) [reply packages/ctl/ctl.js:917]: OK. I’ll replace it next commit.

**Comment by psihius** [comment]
@jake-87 converted it to draft, press ze button when it's ready for review :)

**Comment by jake-87** [comment]
All this needs is a way to refresh the user table so the bulk imports show up, and it's done, as far as I can tell.

**Comment by psihius** [comment]
>

**Comment by psihius** [comment]
> All this needs is a way to refresh the user table so the bulk imports show up, and it's done, as far as I can tell.

You probably need an event/message for that to send out, so the systems grab it from the websocket message bus.
Also, make sure tests don't fail :)

Bonus would be adding tests for this functionality.

**Comment by jake-87** [comment]
Apologies, forgot to run final tests after removing personal stuff. I'll see what I can do in terms of the event.

File: packages/ctl/ctl.js [hunk 2]
```diff
@@ -903,6 +903,26 @@ userCommands.add(new libCommand.Command({
...
+	handler: async function(args, control) {
+		let data = await fs.readFile(args.file, "utf8");
+		let names = data.split("\n");
+		await names.forEach(async (name) => {
+			if (name !== "") {
+				await libLink.messages.setUserAdmin.send(control, {
+					name: name, create: args.create, admin: true,
+				});
+			}
+		});
```
Comment (Hornwitser) [inline packages/ctl/ctl.js:922]: This does not do what you think it does. If you're using a function you're unsure of how it works [consult a reference for it](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach). `forEach` returns `undefined` awaiting it is a logical error here. The callback is not awaited which means all of the requests will be started concurrently and that if any one of them throws Node.js will abort with an unhandled promise rejection. A regular `for (... of ...)` loop would be more appropriate here.

File: packages/ctl/ctl.js [hunk 3]
```diff
@@ -913,16 +913,13 @@ userCommands.add(new libCommand.Command({
 	handler: async function(args, control) {
 		let data = await fs.readFile(args.file, "utf8");
 		let names = data.split("\n");
-		for (let name in names) {
-			if (name in names) {
-				if (names[name] === "") {
-					continue;
-				}
+		await names.forEach(async (name) => {
```
Comment (Danielv123) [inline packages/ctl/ctl.js:916]: This is wrong. `Array.forEach` returns undefined, which isn't a Promise. You want to do `await Promise.all(names.map(async name => ...` instead

Comment (Danielv123) [reply packages/ctl/ctl.js:916]: This goes for all the places you are using forEach here I see. Hornwitser suggested a for loop. The difference is that promise.all executes in parallel while `for` does sequential. I guess for a long whitelist you'd need to go sequential or batched parallel.

Comment (jake-87) [reply packages/ctl/ctl.js:916]: Doing it in parallel was my original intent; unfortunately I did mess up the use of forEach. I’ll correct this next commit.

**Comment by jake-87** [comment]
I am still working on this. Unfortunately the documentation for the link system is less than optimal, and I am having a hard time understanding the code, so I can't give a decent estimate on when the web interface will be finished.

**Comment by psihius** [comment]
@jake-87 ask in Discord? We can also improve the docs this way :)

**Comment by jake-87** [comment]
I'm not going to be able to finish this. Apologies. Feel free to take & merge the CLI code, but the websocket interface is not well documented enough for me to be able to understand it, and while I appreciate the discord's existence, in my mind it is not a good enough replacement for docs.

**Pull request closed without merging**